src/shared/Core/CollectorWriter.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+using KnownGitCfg = GitCredentialManager.Constants.GitConfiguration;
+
+namespace GitCredentialManager
+{
+    public enum CollectorWriterState
+    {
+        Unknown = 0,
+        Stopped = 1,
+        Failing = 2,
+        Healthy = 3,
+    }
+
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class CollectorWriter
```

Comment:
Does this need an explicit `: IDisposable`?

Reply:
Or, as I see later, `DisposableObject`?

Reply:
Yep, it absolutely does. Good catch.